### PR TITLE
add real_path class method for File

### PR DESCRIPTION
**Pull request by jreinert**
open to discussion if this should be called realpath instead like in Ruby

### CODE REVIEW

**Comment by jhass** [comment]
How does this differ from `expand_path`?

**Comment by jreinert** [comment]
it resolves symbolic links

**Comment by jreinert** [comment]
see man 3 realpath

**Comment by jhass** [comment]
I wonder if we can/should combine the two interfaces, something like `File.expand_path(foo, resolve: true)`? In any case adding docs to both that clearly disambiguate them would be great.

**Comment by jreinert** [comment]
hmm. I'm not sure if this should be combined with expand_path. They are really two different things and i think they would be mixed up more easily if combined. But that's just my personal preference.

**Comment by jhass** [comment]
I'm bringing this to the table since I see `real_path`s functionality as a superset of `expand_path`. It does what `expand_path` does and then resolves symlinks and verifies filesystem existence.

**Comment by jreinert** [comment]
i guess so, still: `expand_path` doesn't need any access to the filesystem itself as opposed to `real_path` which implies possible exceptions when called. `expand_path` is "safer" in a sense.

**Comment by ysbaddaden** [comment]
LGTM.

I'd use `realpath` instead of `real_path` thought.

`realpath` is well known (C function, CLI binary) and very different from `expand_path`. The latter is merely a nicer `File.join` that plays with strings, when the former searches for files on the disk, following symlinks and failing when the file doesn't exist.

**Comment by alanwillms** [comment]
Trying to be C-friendly is what ended-up messing with PHP function names. :laughing:

**Comment by ysbaddaden** [comment]
Good point, let's say this is to be ruby compatible? http://ruby-doc.org/core-2.2.2/File.html#method-c-realpath ;-)

**Comment by alanwillms** [comment]
Then :+1: although I suggest there should be a Crystal convention, like "no abbreviations", "every word must be separated by an underscore", etc.

**Comment by jreinert** [comment]
so... realpath for ruby compatibility or real_path for convention and readability?

**Comment by jhass** [comment]
I can't make up my mind, @asterite?

**Comment by jreinert** [comment]
@ysbaddaden ruby compatibility is hard to achieve everywhere anyways. I think like @alanwillms said, there should be a convention. Even better: An official style guide.

**Comment by kirbyfan64** [comment]
I think other languages use `realpath` too, such as [Python](https://docs.python.org/3/library/os.path.html#os.path.realpath) and [Posix](http://man7.org/linux/man-pages/man3/realpath.3.html).

**Comment by asterite** [comment]
@jhass Wait until we discuss this with @waj, we didn't have time to look at this yet

**Comment by will** [comment]
Just a drive-by opinion, but I'd be happy if `#expand_path` did both resolving of symlinks and rooting to `/`. I haven't had a situation that I can remember where I wanted one but __not__ the other. 

That is, there have been times when I've wanted only one of the two, but I wouldn't have cared if the other behavior came along too.

**Comment by jreinert** [comment]
@will you mean `File.real_path("foo/bar/baz") # => "/path/to/foo/bar/baz"`

**Comment by ysbaddaden** [comment]
`expand_path` resolves a path relative to *another* path (doesn't hit the disk); `realpath` eliminates relatives paths in the path *itself*, following all symlinks to eventually point to the actual file that must exist (hits the disk many times).

**Comment by jreinert** [comment]
@asterite @waj you'll have the last say in this. what shall it be?

File: src/file.cr [hunk 1]
```diff
@@ -309,6 +310,12 @@ class File < IO::FileDescriptor
     end
   end
 
+  def self.real_path(path)
+    real_path_ptr = LibC.realpath(path, Pointer(UInt8).null)
```
Comment (asterite) [inline src/file.cr]: You can simply do:

```crystal
real_path_ptr = LibC.realpath(path, nil)
```

**Comment by asterite** [comment]
I prefer `real_path` because it's easier to read and consistent to other names, even though it's incompatible with other languages.

But let me check this with @waj first, and then we'll probably merge this.

**Comment by asterite** [comment]
Well, looks good like this for me. I'd add some docs and use `nil` instead of `Pointer.null`

**Comment by jreinert** [comment]
@asterite i made the suggested changes

**Comment by ysbaddaden** [comment]
Thanks!

**Pull request merged**